File: SpinOperatorLibrary.m
(* ::Package:: *)

(* ::Package:: *)
(**)


BeginPackage["SpinOperatorLibrary`"];

(* Exported symbols *)
Je::usage = "Je[sn, j, m] calculates the matrix element for spin operators.";
Jx::usage = "Jx[n] generates the Jx matrix for dimension n.";
Jx2::usage = "Jx2[n] generates the \!\(\*SuperscriptBox[\(Jx\), \(2\)]\) matrix for dimension n.";
Jy::usage = "Jy[n] generates the Jy matrix for dimension n.";
Jz::usage = "Jz[n] generates the Jz matrix for dimension n.";

Begin["`Private`"]

Je[sn_, j_, m_] := Sqrt[j*(j + 1) - m*(m + sn*1)]
 
Jx[(n_)?IntegerQ] := Module[{diag, b1, b2, m}, b1 = Je[1, n/2, m]/2; 
      SparseArray[{Band[{1, 2}, {n, n + 1}] -> Table[b1, {m, n/2 - 1, -n/2, 
           -1}], Band[{2, 1}, {n + 1, n}] -> Table[Conjugate[b1], 
          {m, n/2 - 1, -n/2, -1}]}, {n + 1, n + 1}]] 
 
Jy[(n_)?IntegerQ] := Module[{diag, b1, b2, m}, b1 = Je[1, n/2, m]/(2*I); 
      SparseArray[{Band[{1, 2}, {n, n + 1}] -> Table[b1, {m, n/2 - 1, -n/2, 
           -1}], Band[{2, 1}, {n + 1, n}] -> Table[Conjugate[b1], 
          {m, n/2 - 1, -n/2, -1}]}, {n + 1, n + 1}]]
 
Jz[(n_)?IntegerQ] := Module[{diag, b1, b2, m}, 
     diag = m; SparseArray[{Band[{1, 1}, {n + 1, n + 1}] -> 
         Table[diag, {m, n/2, -n/2, -1}]}, {n + 1, n + 1}]]
         
Jx2[(n_)?IntegerQ] := Module[{diag,b1,b2,m},
diag=1/4 (Je[-1,n/2,m]Je[1,n/2,m-1]+Je[1,n/2,m]Je[-1,n/2,m+1]);(*diagonal*)
b2=1/4 ( Je[1,n/2,m+1]Je[1,n/2,m]);(*2. upper offdiagonal*)

SparseArray[{
Band[{1,1},{n+1,n+1}]->Table[diag,{m,n/2,-n/2,-1}](*diagonal*),
Band[{1,3},{n-1,n+1}]->Table[b2,{m,n/2-2,-n/2,-1}] (*2. upper offdiagonal*),Band[{3,1},{n+1,n-1}]->Table[b2,{m,n/2-2,-n/2,-1}] (*2. lower offdiagonal*)
},{n+1,n+1}]
]

End[];
EndPackage[];

File: Lipkin model/src/SpinOperatorLibrary.jl
module SpinOperatorLibrary

using LinearAlgebra, SparseArrays

export je, Jx, Jy, Jz, Jx2

"""
    je(sn, j, m)

Calculates the matrix element for spin operators.
"""
function je(sn, j, m)
    return sqrt(j*(j + 1) - m*(m + sn))
end

"""
    Jx(n::Int)

Generates the Jx matrix for a spin of j = n/2 (resulting in an (n+1)×(n+1) matrix).
"""
function Jx(n::Int)
    j = n / 2
    N = n + 1
    row = Int[]
    col = Int[]
    vals = Float64[]
    # Upper diagonal: for i = 1 to n, with m = j - i
    for i in 1:n
        m = j - i
        push!(row, i)
        push!(col, i + 1)
        push!(vals, je(1, j, m) / 2)
    end
    # Lower diagonal: same as above (real numbers, so conjugate is the same)
    for i in 1:n
        m = j - i
        push!(row, i + 1)
        push!(col, i)
        push!(vals, je(1, j, m) / 2)
    end
    return sparse(row, col, vals, N, N)
end

"""
    Jy(n::Int)

Generates the Jy matrix for a spin of j = n/2 (resulting in an (n+1)×(n+1) matrix).
"""
function Jy(n::Int)
    j = n / 2
    N = n + 1
    row = Int[]
    col = Int[]
    vals = ComplexF64[]
    # Upper diagonal: using factor 1/(2im)
    for i in 1:n
        m = j - i
        push!(row, i)
        push!(col, i + 1)
        push!(vals, je(1, j, m) / (2im))
    end
    # Lower diagonal: note that the complex conjugate of 1/(2im) is -1/(2im)
    for i in 1:n
        m = j - i
        push!(row, i + 1)
        push!(col, i)
        push!(vals, -je(1, j, m) / (2im))
    end
    return sparse(row, col, vals, N, N)
end

"""
    Jz(n::Int)

Generates the Jz matrix (diagonal) for a spin of j = n/2 (resulting in an (n+1)×(n+1) matrix).
"""
function Jz(n::Int)
    j = n / 2
    N = n + 1
    diag = [j - (i - 1) for i in 1:N]
    return spdiagm(0 => diag)
end

end # module
